src/range_inclusive.rs

```diff
@@ -147,6 +148,93 @@ parallel_range_impl! {i64}
 parallel_range_impl! {u128}
 parallel_range_impl! {i128}
 
+// char is special
+impl ParallelIterator for Iter<char> {
+    type Item = char;
+
+    fn drive_unindexed<C>(self, consumer: C) -> C::Result
+    where
+        C: UnindexedConsumer<Self::Item>,
+    {
+        self.drive(consumer)
+    }
+
+    fn opt_len(&self) -> Option<usize> {
+        Some(self.len())
+    }
+}
+
+// Range<u32> is broken on 16 bit platforms, may as well benefit from it
+impl IndexedParallelIterator for Iter<char> {
+    // Split at the surrogate range first if we're allowed to
+    fn drive<C>(self, consumer: C) -> C::Result
+    where
+        C: Consumer<Self::Item>,
+    {
+        if let Some((start, end)) = self.bounds() {
```

Comment:
Hmm, `bounds()` won't work until Rust 1.45...

FWIW, I was also looking at that function in godbolt, and it works a lot better for `char` this way:

```rust
    if range.size_hint().0 > 0 {
        Some((*range.start(), *range.end()))
    } else {
        None
    }
```
https://rust.godbolt.org/z/8FrRRW

But it does require that we rely on `size_hint()` to be correct (like `TrustedLen`). I guess this code shouldn't be a hotspot anyway, so it probably doesn't matter.

Reply:
The correct way to do it would be to use `.is_empty()` if/when it stabilizes.

That can be optimized in a followup if necessary/desired, I suppose. I don't think it would be _incorrect_ to use `size_hint` here, so long as there's a note about relying on `TrustedLen`-like semantics for correctness (but hopefully not safety?).

Reply:
Side note: oof, there's a unwrap/expect panic still in there. I guess I need to go tweak the impls again to see if I can remove that (as it _should_ be completely unreachable, and if it isn't, that's definitely a bug.)

Reply:
Yep, the generated asm is a _lot_ better when I add in the missed `_unchecked` to `next_back`: https://rust.godbolt.org/z/9dviLq

We already have it for `next` because I spotted the missed optimization out of the unreachable panicking code, but for some reason I didn't think to check reverse iteration for the same thing.

Reply:
> I don't think it would be _incorrect_ to use `size_hint` here, so long as there's a note about relying on `TrustedLen`-like semantics for correctness (but hopefully not safety?).

There's no safety problem, just `start()` and `end()` are _unspecified_ when empty.